torchtune/models/llama2/_tokenizer.py

```diff
@@ -0,0 +1,184 @@
+
+from torchtune.data import Message, truncate
+from torchtune.data.tokenizers import SentencePieceEncoding, Tokenizer
+
+WHITESPACE_CHARS = [" ", "\n", "\t", "\r", "\v"]
+
+
+class Llama2Tokenizer(Tokenizer):
+    """
+    Llama2's implementation of the SentencePiece tokenizer. Llama2Tokenizer does
+    not include any additional special tokens. The prompt template described in
+    https://llama.meta.com/docs/model-cards-and-prompt-formats/meta-llama-2/ describes
+    [INST][/INST] and <<SYS>><</SYS>> as special tokens but these are not registered
+    as unique ids and are tokenized as normal text. When using this tokenizer on the
+    pre-trained model for inference, it is strongly encouraged to apply the
+    :class:`~torchtune.data.Llama2ChatFormat` to your data beforehand to add the
+    [INST] and <<SYS>> for optimal performance. For fine-tuning, this is not required.
+    For more details, see https://pytorch.org/torchtune/main/tutorials/chat.html#tokenizing-prompt-templates-special-tokens.
+
+    Args:
+        path (str): Path to pretrained SentencePiece tokenizer file.
+
+    Example:
+        # Accepts only non-batched input for now
+        >>> tokenizer = Llama2Tokenizer("/path/to/spm_model")
+        >>> tokenized_text = tokenizer.encode("Hello world!", add_bos=True, add_eos=True)
+        >>> print(tokenized_text)
+        [1, 31587, 29644, 102, 2]
+    """
+
+    def __init__(
+        self,
+        path: str,
+    ):
+        self._spm_model = SentencePieceEncoding(path)
+
+        # Original tokenizer has no pad_id, which causes indexing errors when batch training
+        self._spm_model.pad_id = 0
+
+        # This is used in tokenize_messages: if the tokenizer does not
+        # encode whitespace, then we can more easily split strings
+        # on whitespace characters and encode them separately.
+        self.encodes_whitespace = any(
+            [self._spm_model.encode(c) for c in WHITESPACE_CHARS]
+        )
+
+        # During generation, stop when eos_id is encountered
+        self.stop_tokens = [self.eos_id]
+
+    @property
+    def eos_id(self):
+        return self._spm_model.eos_id
+
+    @property
+    def bos_id(self):
+        return self._spm_model.bos_id
+
+    @property
+    def pad_id(self):
+        return self._spm_model.pad_id
+
+    @property
+    def vocab_size(self):
+        return self._spm_model.vocab_size
+
+    def encode(
+        self,
+        text: str,
+        add_bos: bool = True,
+        add_eos: bool = True,
+        trim_leading_whitespace: bool = False,
+    ) -> List[int]:
+        return self._spm_model.encode(
+            text,
+            add_bos=add_bos,
+            add_eos=add_eos,
+            trim_leading_whitespace=trim_leading_whitespace,
+        )
+
+    def decode(
+        self,
+        token_ids: List[int],
+        include_special: bool = False,
+    ) -> str:
+        return self._spm_model.decode(token_ids, include_special=include_special)
```

Comment:
Yeah I mainly surfaced the encode/decode in the model tokenizers for generation utilities. Otherwise we'd have to access the underlying model for encoding/decoding... which maybe is ok if we enforce a common class attribute name in the model tokenizer protocol.

Reply:
Could we call it "decode_messages" on the `ModelTokenizer` and have it handle filtering out the special tokens?

Reply:
BUmp

Reply:
sorry I missed this. Yes I agree, will do that

Reply:
actually, I removed `include_special` from decode here, so this no longer applies right? and the generate recipe still requires using `decode` from the `ModelTokenizer`